### PR TITLE
RFC: Widgets / Plugins / Middleware / Argument Injector thingie

### PR DESCRIPTION
**Pull request by przemyslaw-przylucki**
Good morning!

I've been tinkering with the idea from the initial PR a bit more and came up with this initial concept. 

**Note:** I haven't spent much time on this, let's treat this as a draft of functionality that we want to have.

So, as the title suggests - not sure about name, I'll just stick with widgets for now which is probably the worst name but 🤷 

The goal of widgets is to create a way to inject additional arguments into the console. 
This can be useful for example when our command requires an API key, that could be automatically injected if not provided, or it'd use the passed one.

Widgets can be defined both globally (maybe it'd be nice to add "namespace" param for the `#[GlobalConsoleWidget]` so all commands within i.e `db:` namespace would be affected)  and locally in command definition. 

I think the biggest place for discussion is what contract the invoke method should have and what should be modifiable - but this would require validation to run twice:
1. after resolving the passed arguments
2. after running invoke on all widgets

### CODE REVIEW

File: src/Widgets/GlobalConsoleWidget.php [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Console\Widgets;
+
+use Attribute;
+
+#[Attribute(Attribute::TARGET_CLASS)]
```
Comment (przemyslaw-przylucki) [inline src/Widgets/GlobalConsoleWidget.php:9]: These globals should probably never be used for anything other than --help, --verbosity --silent etc.

File: src/ConsoleInputBuilder.php [hunk 2]
```diff
@@ -4,13 +4,18 @@
 
 namespace Tempest\Console;
 
+use ReflectionClass;
```
Comment (przemyslaw-przylucki) [inline src/ConsoleInputBuilder.php:7]: this obviously needs a refactor before it gets merged =)

File: src/ConsoleArgumentDefinition.php [hunk 3]
```diff
@@ -14,14 +15,15 @@ public function __construct(
...
+        public ?int $position,
         public ?string $description = null,
         public array $aliases = [],
         public ?string $help = null,
+        public bool $isPositional = true,
     ) {
     }
 
-    public static function fromParameter(ReflectionParameter $parameter): ConsoleArgumentDefinition
+    public static function fromParameter(ReflectionParameter $parameter, bool $asPositional = true): ConsoleArgumentDefinition
```
Comment (przemyslaw-przylucki) [inline src/ConsoleArgumentDefinition.php:26]: Injected arguments always have to be named arguments to avoid collisions

**Comment by brendt** [comment]
I think there's still room for improvement to what I did, but maybe it's best to take more incremental steps. 

In truth, I haven't been able to look at this PR in depth. Maybe keep it open for now, I'll definitely go through it and might cherry-pick some stuff straight into main when #21 is merged

**Comment by brendt** [comment]
So, I really like the widget approach, but I feel like we need to get middleware straight first. I'll take care of that, and then we can revisit this idea

**Pull request closed without merging**